### ISSUE TITLE
[FR] Windows support: Support pageant protocol

### ISSUE THREAD

**Issue by pyhedgehog** [ronf/asyncssh#69]
Putty is de-facto standard ssh client for windows OS-family so support for Pageant (key agent for putty) protocol will be very useful. There are also some alternative client _and_ agent implementations (most notably KeeAgent - addon for KeePass).
Reference links:
 * https://git.tartarus.org/?p=simon/putty.git;a=blob;f=ssh.c;h=b09c82a4d6bcfee4ac52092b3e717d816023910f;hb=HEAD#l4352
 * https://git.tartarus.org/?p=simon/putty.git;a=blob;f=windows/winpgntc.c
 * https://github.com/cuviper/ssh-pageant
 * https://github.com/dlech/KeeAgent/blob/master/KeeAgent/KeeAgentExt.cs
 * https://github.com/dlech/SshAgentLib/blob/master/SshAgentLib/PageantClient.cs
 * https://github.com/wesleyd/charade

**Comment by ronf** [comment]
Thanks for the links here. While there doesn't seem to be a protocol doc for Pageant, it appears that it may be compatible with the SSH agent protocol in terms of the messages sent back and forth, with the difference being the use of a memory-mapped file and Windows message passing to actually send the request and get back the response.

I'm not a Windows programmer, but with a bit of fiddling around I was able to write some Python code that successfully sent a request to Pageant. It was surprisingly simple -- only around a dozen lines of code! It will introduce a dependency on PyWin32, though -- specifically win32api, win32con, and win32ui, as well as ctypes and ctypes.wintypes.

I've got a few other work items I'm already in the middle of, but I'll add this to the list!

**Comment by ronf** [comment]
I just checked in a first cut at adding support for Pageant into the "develop" branch. It should now be possible for AsyncSSH to access authentication keys stored in Pageant when it is making outbound SSH connections on Windows.

If you'd like to give this a try, you can install the develop branch of AsyncSSH using the command:

    pip install git+https://github.com/ronf/asyncssh@develop

Please let me know if you run into any issues. If all goes well, I'd like to release this in the upcoming 1.8.0 release in the next week or two.

**Comment by pyhedgehog** [comment]
Cool! After `pip install pypiwin32` it works!

**Comment by ronf** [comment]
Great! Thanks for testing it, and for mentioning the specific PyPI package you needed. I did the testing here with ActiveState Python, and so the PyWin32 packages I needed were already installed. I'll add a note in the documentation about the "pypiwin32" dependency, though, like I did for "bcrypt" and "libnacl".

**Comment by pyhedgehog** [comment]
It would be very useful to add some logging to agent.py:
```python
except ImportError as err: # pragma: no cover
    def open_agent(loop, agent_path, *, reason=err):
        """Dummy function if we're unable to import agent support"""

        # pylint: disable=unused-argument

        raise OSError(errno.ENOENT, 'Agent support not available: '+str(reason))
...
def connect_agent...
...
    except (OSError, ChannelOpenError) as err:
        logger.debug('connect_agent failed: %s', err)
```

**Comment by ronf** [comment]
Done - thanks for the suggestion.

I've been meaning to add debug logging like this elsewhere in AsyncSSH, but haven't gotten around to it. I have to start somewhere, though, and this is as good a place as any!

**Comment by ronf** [comment]
This support is now released in AsyncSSH 1.8.0. Thanks for the suggestion and for the details you provided about how to implement it!

**Comment by bms** [comment]
This is awesome. Pageant is go-to for GUI mode, but unfortunately the MSYS/Cygwin ecosystem has other ideas; it wants to use 'classic' ssh-agent. Enter this bridge: https://github.com/cuviper/ssh-pageant

It may also be worth trying this stuff with PKCS#11 tokens.

**Comment by ronf** [comment]
Thanks, Bruce! It been years since I've played with Cygwin, but I managed to install it on a Windows VM tonight and it looks like ssh-pageant already works correctly with the SSH agent code in AsyncSSH that was written for UNIX! If you have Pageant and ssh-pageant running and you set SSH_AUTH_SOCK in your Cygwin shell as described in the ssh-pageant docs, AsyncSSH will use the keys managed by Pageant. I was even able to get SSH agent forwarding to work, and could access the Pageant keys remotely from the machine I ssh'd into.

Were you also able to get it working?
